### PR TITLE
Fix warnings

### PR DESCRIPTION
**Pull request by SkySkimmer**
This fixes warnings `+a-4-9-27-41-42-44-45-48-50` and adds configure options for those who want to enable that set (the default has not changed).
I'm not sure what warning 29 (unescaped end of line) is about so maybe that commit should be removed or isn't right.
Micromega had a surprising amount of unused values, I'm not sure if that's normal.

### CODE REVIEW

File: kernel/nativecode.ml
Comment (SkySkimmer) [inline kernel/nativecode.ml]: 37 is unused constructor, probably manipulated where ocaml can't see.
32 is unused value, I'm less sure that this one is spurious.

File: kernel/nativevalues.ml
Comment (SkySkimmer) [inline kernel/nativevalues.ml]: I don't know why this type is unused.

**Comment by ejgallego** [comment]
Cool! You may want to update the .merlin file too.

**Comment by silene** [comment]
Commit dbf3f00 is plain wrong. Indeed,

    "A
        B"

is parsed as `"A\n    B"` (for some reason, github is eating my spaces, there are four of them here), while

    "A\
        B"

is parsed as `"AB"`. Notice how you lost both `\n` and the indentation. So using `\` is especially wrong for usage strings and for `coq_makefile`. But this also shows a place where the current code was broken.

**Comment by SkySkimmer** [comment]
So it should be 
```
"foo\
\nbar"
```
as in the usage.ml rather than 
```
(* warns *)
"foo
bar"
```
as in coq_makefile's usage string?

**Comment by silene** [comment]
Yes, putting `\n` in front of a line is damn ugly, but it is the only way I know of to ensure that indentation is not discarded by the compiler.

**Comment by SkySkimmer** [comment]
Fixed the newlines.
I tried to add a travis target which runs the tests but couldn't figure it out and I'm tired of spamming. My last attempt is at https://github.com/SkySkimmer/coq/commit/b66b020920e5a2147425bc6286d4307ae5cf3c93 if anyone wants to look.

**Comment by ejgallego** [comment]
I am not sure the set of enabled warnings should be a configure option. IMHO this usually leads to ineffective warnings as people will just not enable them. In my opinion strengthening the set of defaults warnings is OK.

**Comment by SkySkimmer** [comment]
Sounds good to me.
If someone wants a warning disabled they can say so here, or if they encounter it after this PR is merged they can disable it at that point and make it part of their own PR.

**Comment by ejgallego** [comment]
Sounds good, let's enable them by default then !! We can always refine in the beta phase.

**Comment by ejgallego** [comment]
Indeed the biggest problem with warnings is that different compiler versions may produce slightly different results. Our testing should help with that, but IMHO we will need to place this in the wild and react to the real world until the set of warnings is refined into a workable subset.

**Comment by SkySkimmer** [comment]
Had to rebase on trunk since #538 left an unused open.

**Comment by SkySkimmer** [comment]
I'm making the deprecated warning into a non error since otherwise 4.04 deprecates things like Char.uppercase in favor of uppercase_ascii which doesn't exist yet for 4.02.3.

File: configure.ml
Comment (Zimmi48) [inline configure.ml]: Does the `-warn-more` mode exist or are you talking about the state where `-no-warn-error` option is not passed?

**Review by Zimmi48 — COMMENTED** [review]
Do I understand correctly that following @ejgallego's suggestion, warnings are turned into errors by default? It feels a bit radical and dangerous to me because we won't test all compiler versions. Also it will encourage developers to just pass the `no-warn-error` option to not be bothered in the middle of a development. That no new warning is introduced should be checked in a clean-up phase only. So why not just add a Travis target to test for warnings as was the original suggestion?

File: configure.ml
Comment (SkySkimmer) [inline configure.ml]: oops

**Comment by SkySkimmer** [comment]
>Do I understand correctly that following @ejgallego's suggestion, warnings are turned into errors by default?

yes
>It feels a bit radical and dangerous to me because we won't test all compiler versions.

I'm not sure what you mean by that.
>Also it will encourage developers to just pass the no-warn-error option to not be bothered in the middle of a development.

That's not a problem, is it?
>That no new warning is introduced should be checked in a clean-up phase only.

I'm not sure what you mean by "clean-up phase".
>So why not just add a Travis target to test for warnings as was the original suggestion?

I think even if -no-warn-error ends up the default it would be better to just add -warn-error to the already present travis targets.

**Comment by Zimmi48** [comment]
> > It feels a bit radical and dangerous to me because we won't test all compiler versions.
>
> I'm not sure what you mean by that.

I mean I wouldn't like the compilation of Coq to fail on some user's machine because of a stupid warning.

> > Also it will encourage developers to just pass the `no-warn-error` option to not be bothered in the middle of a development.
>
> That's not a problem, is it?

If the developers deactivate the mode and the users are just bothered by it, what is the point of having it turn on by default?

> I think even if `-no-warn-error` ends up the default it would be better to just add `-warn-error` to the already present travis targets.

+1

**Comment by ejgallego** [comment]
I see @Zimmi48's point, I think `-warn-on-error` should be the default for Travis and development builds, but indeed user tarballs and OPAM should disable this flag.

**Comment by ejgallego** [comment]
I think having it as a default for the dev tree makes sense as most developers will just forget to pass -warn-on-error

**Comment by Zimmi48** [comment]
> I see @Zimmi48's point, I think -warn-on-error should be the default for Travis and development builds, but indeed user tarballs and OPAM should disable this flag.

Today, lots of users do `git clone` instead of downloading a tarball.

> I think having it as a default for the dev tree makes sense as most developers will just forget to pass -warn-on-error

Yeah but developers sometimes forget to test the test-suite as well. This is what Travis is there for...

**Comment by SkySkimmer** [comment]
configure.ml can't figure out if someone is a user or a developer so it makes sense to have -warn-error be an explicit choice.
Now I'm enabling warnings but not warn-error by default, and travis enables warn-error.
There are a bunch of builds waiting due to recent merges so it will take some time until travis gets to this PR.

**Comment by SkySkimmer** [comment]
I'm not sure if it's a problem, but merges from 8.6 into trunk seem like an easy way to introduce warnings without being detected.

**Comment by Zimmi48** [comment]
> I'm not sure if it's a problem, but merges from 8.6 into trunk seem like an easy way to introduce warnings without being detected.

Well, no. Travis also tests direct commits to trunk. There is even a Travis badge in the README now. So if such a merge introduced a warning. This badge would turn from "Build passing" to "Build failed".

**Comment by ejgallego** [comment]
> configure.ml can't figure out if someone is a user or a developer so it makes sense to have -warn-error be an explicit choice.

Once the version system is in place indeed it would be able to detect if it is a clone of a git tree or a tarball. IMHO people cloning the tree can bear with tweaking configure; of course `warn-on-error` is easy to tweak later, but I am not comfortable with the different modes on Travis and dev, I would rather have them sync'ed.

In a sense this is a deeper question, what is the goal of CI? So CI is for testing that builds are reproducible, and general sanity checks, but IMHO assigning the CI the role of the warnings curator is the wrong one.

**Comment by ejgallego** [comment]
> I'm not sure if it's a problem, but merges from 8.6 into trunk seem like an easy way to introduce warnings without being detected.

This is IMHO another argument to support a default `error-on-warn`. The release managers build-check the merges before pushing, so it makes sense to make them aware of new warnings.

**Comment by Zimmi48** [comment]
@ejgallego We do already detect in the configure file if we are in a git tree (cf. https://github.com/Zimmi48/coq/blob/trunk/configure.ml#L431).

**Comment by ejgallego** [comment]
Indeed we could use this and enable warn-on-error. Maybe I am too paranoid but I feel a pile of problems if Travis has it enabled and our normal workflow doesn't. I hope we can do better once we make more progress on #483

**Comment by ejgallego** [comment]
But I fully agree with your concerns on `-warn-on-error`

**Comment by Zimmi48** [comment]
I feel that indeed you are a bit paranoid ;-) but my concern was a minor one so I won't block merging of this PR whatever the default is chosen to be.

**Comment by ejgallego** [comment]
It is more of a "state problem" I think. I feel like my head can only hold at most 2 bits of state, so now adding another bit "is -warn-on-error" enabled?" when I need to diagnose travis failures tends to overload my stack, but dunno, I'd love to hear what others think.

**Comment by SkySkimmer** [comment]
Compiling with OCaml 4.04 is failing due to warning 52. Some of them are internal to Coq and I'll replace them with specific exceptions, but I'm not sure what to do about these:
- matching `Sys_error "Is a directory"` to print a custom error in coqwc https://github.com/coq/coq/blob/trunk/tools/coqwc.mll#L250, git blame says it was there from the start in 9dd608e78fb7d392ea88b47f396f5077021337d5 (@backtracking)
- matching `Invalid_argument "equal: functional value"` in https://github.com/coq/coq/blob/trunk/lib/cErrors.ml#L128 to make that exception critical (for canaries), introduced 2a0d260c9c80c07844605fcb6844bb9cfdfeb0fd (@ppedrot)

There are basically 2 solutions:
- locally disable the warning
- remove the custom behaviour, ie say "system error: is a directory" instead of "system error: foo is a directory" and make all `Invalid_argument`s critical errors.

(Compilation also fails due to new merges in trunk, and in 4.04 due to some unused modules like Hsorts in constr.ml, but those are simple)

**Comment by JasonGross** [comment]
Note that if warnings are errors by default, bisecting Coq will require disabling this behavior as we move to newer compilers.

File: parsing/pcoq.ml
Comment (SkySkimmer) [inline parsing/pcoq.ml]: 4.04 seems more sensitive for preexisting warnings, like this unused open.

**Comment by ejgallego** [comment]
@SkySkimmer indeed warning 52 is a pain.

I suggest we try to have a quick course of action with this patch because we seems to be losing loads of time blocking and rebasing PRs instead of doing actual development.

My proposal is then:
- Let's merge this patch now, with `-warn-on-error` disabled.
- Let's keep Travis building things with `-warn-on-error` disabled by default.
- Let's postpone the discussion on whether `-warn-on-error` should be enabled by default to a WG.

File: vernac/record.ml
Comment (mattam82) [inline vernac/record.ml]: This should rather be used in the code below, I think that's an overlook

File: kernel/names.ml
Comment (mattam82) [inline kernel/names.ml]: What's the warning here, out of curiosity?

**Review by mattam82 — APPROVED** [review]
I didn't catch any more issues, I agree with @ejgallego that we should rather quickly merge this and leave the warn-error decision for later.

File: kernel/names.ml
Comment (SkySkimmer) [inline kernel/names.ml]: Unused type t
Looking at it more I should just remove Projection.SyntacticOrd.t since it doesn't appear in the mli and is the same as Projection.t

**Review by ejgallego — APPROVED** [review]

**Review by ejgallego — CHANGES_REQUESTED** [review]
Please make this match the default configuration

**Comment by SkySkimmer** [comment]
@ejgallego 
>Please make this match the default configuration

What do you mean by "this"? (I think your message was supposed to be attached to some part of the diff?)

**Comment by ejgallego** [comment]
@SkySkimmer yup sorry I messed up.

File: .travis.yml
Comment (ejgallego) [inline .travis.yml]: Please make the `-warn-error` setting match the default configuration. You could use `EXTRA_CONF` if you wanna create an optional allow fail target, but I suggest we tweak travis in a new PR, we could also cleanup the IDE building by using YAML refs.

**Comment by SkySkimmer** [comment]
- disabled -warn-error in travis (ie left default)
- used `warn_cannot_define_projection` in vernac/record.ml instead of deleting it
- removed type copies in kernel/names.ml SyntacticOrd submodules

**Review by ejgallego — APPROVED** [review]

**Comment by ejgallego** [comment]
Great, thanks! LGTM.

**Comment by SkySkimmer** [comment]
Travis passed!

File: engine/evd.ml
Comment (maximedenes) [inline engine/evd.ml]: Should this go away?

File: kernel/constr.ml
Comment (maximedenes) [inline kernel/constr.ml]: @mattam82 was it expected that this was disabled?

File: plugins/ltac/tauto.ml
Comment (maximedenes) [inline plugins/ltac/tauto.ml]: Is this unused? If yes, why do we keep it?

File: printing/prettyp.ml
Comment (maximedenes) [inline printing/prettyp.ml]: Should we remove this?

File: kernel/constr.ml
Comment (SkySkimmer) [inline kernel/constr.ml]: It's in sorts.ml

File: plugins/ltac/tauto.ml
Comment (herbelin) [inline plugins/ltac/tauto.ml]: Personally, it is relevant to me to remember how it worked in 8.4. And having it type-checked by ocaml ensures a better correctness than if it were a comment. But if it is cluttering too much for others, I can also live with `git log`.

Comment (SkySkimmer) [inline plugins/ltac/tauto.ml]: It was introduced in e1f5180e88bf02a22c954ddbdcbdfeb168d264a6 (@herbelin) and was already not used at that time.
I don't know.

File: engine/evd.ml
Comment (SkySkimmer) [inline engine/evd.ml]: I think it doesn't look bad next to the `NamedDecl` but it could go away if someone wants it to.

File: plugins/ltac/tauto.ml
Comment (maximedenes) [inline plugins/ltac/tauto.ml]: @herbelin I see, thanks for the explanation! If it is an 8.4 setting, maybe 8.7 is the right time to remove it, I guess.

**Comment by SkySkimmer** [comment]
@maximedenes so do I remove the things you commented on?

**Comment by maximedenes** [comment]
I think it can be merged as it is, we'll remove that later if we want to. I'm having a last look at the native_compiler ones.

**Comment by herbelin** [comment]
I just found a problem with the `[@@@ocaml.warning "-32"]` in trace mode of `coqtop.byte`. Probably because of camlp5 which does not support it (yet?), the standard debugging protocol `Drop. #use"include";;` fails with a syntax error. Is it ok if we (temporarily) remove this attribute from the file `top_printers.ml`?

**Comment by SkySkimmer** [comment]
I'll make a PR to remove it permanently, putting it there was a mistake (I saw the unused value warning and thought it was because of some strange debug thing I didn't know about, but in fact it's because a couple lets are shadowed).

**Comment by SkySkimmer** [comment]
See #619.

**Comment by herbelin** [comment]
@SkySkimmer: Great!

**Pull request merged**
